src/rewriter/basic_rewrite_rcons.cpp

```diff
@@ -201,8 +216,92 @@ bool BasicRewriteRCons::ensureProofMacroBoolNnfNorm(
+  Node lhs = eq[0];
+  Assert(lhs.getKind() == Kind::STRING_SUBSTR);
+  theory::strings::Rewrite rule;
+  // call the same utility that proved it
+  theory::strings::ArithEntail ae(nullptr);
+  theory::strings::StringsEntail sent(nullptr, ae, nullptr);
+  std::vector<Node> ch1;
+  std::vector<Node> ch2;
+  Node rhs = sent.rewriteViaMacroSubstrStripSymLength(lhs, rule, ch1, ch2);
+  Trace("brc-macro") << "...was via string rewrite rule " << rule << std::endl;
+  Assert(rhs == eq[1]);
+  TypeNode stype = lhs.getType();
+  Node cm1 = theory::strings::utils::mkConcat(ch1, stype);
+  Node cm2 = theory::strings::utils::mkConcat(ch2, stype);
+  Node cm;
+  // depending on the rule, are either stripping from front or back
+  if (rule == theory::strings::Rewrite::SS_STRIP_END_PT)
+  {
+    cm = nm->mkNode(Kind::STRING_CONCAT, cm1, cm2);
+  }
+  else
+  {
+    cm = nm->mkNode(Kind::STRING_CONCAT, cm2, cm1);
+  }
+  if (cm == lhs[0])
+  {
+    return false;
+  }
+  Node eq1 = nm->mkNode(Kind::EQUAL, lhs[0], cm);
+  // It is likely provable by ACI_NORM. However, if it involves splitting
+  // word constants, we require going through the rewrite term converter.
+  if (expr::isACINorm(lhs[0], cm))
+  {
+    cdp->addStep(eq1, ProofRule::ACI_NORM, {}, {eq1});
+    Trace("brc-macro") << "- via ACI_NORM " << eq1 << std::endl;
+  }
+  else
+  {
+    //             ----------------- via encode transform
+    //             (t = s) = (r = r)
+    // ----- REFL  ------------------ SYMM
+    // r = r       (r = r) = (t = s)
+    // ---------------------------------- EQ_RESOLVE
+    // t = s
+    RewriteDbNodeConverter rdnc(nodeManager());
+    Node eq1t = rdnc.convert(eq1);
+    Assert(eq1t.getKind() == Kind::EQUAL);
+    if (eq1t[0] != eq1t[1])
+    {
+      return false;
+    }
+    Node equiv = eq1.eqNode(eq1t);
+    ensureProofForEncodeTransform(cdp, eq1, eq1t);
+    Node equivs = eq1t.eqNode(eq1);
+    cdp->addStep(equivs, ProofRule::SYMM, {equiv}, {});
+    cdp->addStep(eq1t, ProofRule::REFL, {}, {eq1t[0]});
+    cdp->addStep(eq1, ProofRule::EQ_RESOLVE, {eq1t, equivs}, {});
+  }
+  std::vector<Node> cargs;
+  ProofRule cr = expr::getCongRule(lhs, cargs);
+  Node eq2 = lhs[1].eqNode(lhs[1]);
+  Node eq3 = lhs[2].eqNode(lhs[2]);
+  cdp->addStep(eq2, ProofRule::REFL, {}, {lhs[1]});
+  cdp->addStep(eq3, ProofRule::REFL, {}, {lhs[2]});
+  Node lhsm = nm->mkNode(Kind::STRING_SUBSTR, cm, lhs[1], lhs[2]);
+  Node eqLhs = lhs.eqNode(lhsm);
+  cdp->addStep(eqLhs, cr, {eq1, eq2, eq3}, cargs);
+  Node eqm = lhsm.eqNode(rhs);
+  // Note that this is not marked simple, since it may require length
+  // entailment to prove.
+  // Note that there are three cases of string rewrites handled by this macro,
+  // where we expect this trusted step to be filled by one of 3 RARE rewrites,
+  // namely:
+  // str-substr-len-include, str-substr-len-include-pre, str-substr-len-skip.
+  cdp->addTrustedStep(eqm, TrustId::MACRO_THEORY_REWRITE_RCONS, {}, {});
```

Comment:
Note: this search may be an overkill given that it's easy to determine which rule to apply.

Reply:
The problem is we don't have access to those rules here, since they are provided on-demand in RARE.

This is a bit of a weird dependency where the core of cvc5 relies on an on-demand RARE rule (the same was true for the bool-nnf elaboration).

Reply:
> The problem is we don't have access to those rules here, since they are provided on-demand in RARE.

I thought we do? Or do you not want to depend on generated RARE Enums?

> the same was true for the bool-nnf elaboration

This one is fine because it's a pain to figure out which rule to use.

Reply:
It's unclear to me if we should. 

Nevertheless, rebooting the search here is the most flexible solution.